pkg/iptables/iptables.go

```diff
@@ -8,26 +8,61 @@ import (
 	"strings"
 
 	"github.com/grafana/xk6-disruptor/pkg/agent/protocol"
+	"github.com/grafana/xk6-disruptor/pkg/iproute"
 	"github.com/grafana/xk6-disruptor/pkg/runtime"
 )
 
-const redirectCommand = "%s PREROUTING -t nat -i %s -p tcp --dport %d -j REDIRECT --to-port %d"
-
-const resetCommand = "%s INPUT -i %s -p tcp --dport %d -m state --state ESTABLISHED -j REJECT --reject-with tcp-reset"
+// redirectRule is a netfilter rule that redirects traffic destined to the application's port to the proxy.
+// As per https://unix.stackexchange.com/a/112232/39698, traffic coming from non-local addresses cannot be redirected
+// to local addresses, so the proxy is required to listen on, in addition to localhost, the same address where the
+// application also listens on (typically `eth0`).
+// As per https://upload.wikimedia.org/wikipedia/commons/3/37/Netfilter-packet-flow.svg, locally originated traffic,
+// such as port-forwarded or traffic generated from inside the container, does not traverse PREROUTING. Likewise,
+// traffic coming from the outside of the pod does not traverse OUTPUT, so this rule has to be instantiated for both
+// OUTPUT and PREROUTING chains.
+//
+// As a technicality, the PREROUTING rule does not actually need `! -s %s/32`, as traffic from the proxy to upstream
+// does not traverse PREROUTING but OUTPUT. We keep it anyway for consistency.
+const redirectRule = "%s %s -t nat " + // For traffic traversing the nat table
+	"-p tcp --dport %s " + // Sent to the upstream application's port
+	"! -s %s/32 " + // Unless it is sent from the proxy IP address
+	"-j REDIRECT --to-port %s" // Forward it to the proxy address
+
+// redirectChains are the chains in which the redirectRule should be added. see redirectRule for details.
+//
+//nolint:gochecknoglobals // Read-only constant-ish.
+var redirectChains = []string{"OUTPUT", "PREROUTING"}
+
+// resetCommand creates an iptables rule that closes existing and new connections to the upstream application's port,
+// forcing clients to re-establish connections.
+const resetCommand = "%s INPUT " + // For traffic traversing the INPUT chain
+	"! -s %s/32 " + // Not coming from the proxy address
+	"! -d %s/32 " + // Not directed to the proxy address
```

Comment:
This part of the rule is redundant, as the proxy listens on all interfaces. I'll garbage-collect this.